### PR TITLE
Implement support for optimistic block finality

### PR DESCRIPTION
**Pull request by kobayurii**
In this pull request, we introduce support optimistic blocks. 

Recently we've added additional metrics to measure the number of proxy requests from the ReadRPC to the native one. Additionally, we have a metric that counts how many of the requests are for the blocks with optimistic finality. This finality is not yet supported in the ReadRPC. The metrics revealed that a significant portion of the requests are for the optimistic blocks - 85% of all requests.

Implementation of `optimistic_blocks` support will help to reduce the number of proxied requests and help to optimize and speed up the service.

### CODE REVIEW

**Comment by kobayurii** [comment]
https://github.com/near/read-rpc/issues/186

File: rpc-server/src/modules/blocks/mod.rs [hunk 1]
```diff
@@ -28,9 +30,237 @@ impl From<near_primitives::views::BlockView> for CacheBlock {
...
+    pub stream_message: near_indexer_primitives::StreamerMessage,
+}
+
+impl BlockInfo {
+    pub async fn new_from_block_view(block_view: near_primitives::views::BlockView) -> Self {
+        Self {
+            block_cache: CacheBlock::from(&block_view),
+            stream_message: near_indexer_primitives::StreamerMessage {
+                block: block_view,
+                shards: vec![],
```
Comment (khorolets) [inline rpc-server/src/modules/blocks/mod.rs:45]: Please leave a comment explaining why it is empty here.

File: rpc-server/src/modules/blocks/mod.rs [hunk 2]
```diff
@@ -28,9 +30,237 @@ impl From<near_primitives::views::BlockView> for CacheBlock {
...
+
+    pub async fn block_view(&self) -> near_primitives::views::BlockView {
+        self.stream_message.block.clone()
+    }
+
+    pub async fn shards(&self) -> Vec<near_indexer_primitives::IndexerShard> {
+        self.stream_message.shards.clone()
+    }
+
+    pub async fn account_changes_in_block(
```
Comment (khorolets) [inline rpc-server/src/modules/blocks/mod.rs:67]: Please leave doc-strings explaining the reasoning behind these methods here.

File: rpc-server/src/modules/queries/methods.rs [hunk 3]
```diff
@@ -241,26 +241,50 @@ async fn view_account(
         account_id,
         block.block_height
     );
-
-    let account = data
-        .db_manager
-        .get_account(account_id, block.block_height)
+    if let Ok(result) = data
```
Comment (khorolets) [inline rpc-server/src/modules/queries/methods.rs:244]: Put a comment explaining the logic of relying on the Err for non-optimistic queries for future us.

Comment (khorolets) [reply rpc-server/src/modules/queries/methods.rs:244]: Or break down so different functions as discussed on the call

File: rpc-server/src/modules/queries/methods.rs [hunk 4]
```diff
@@ -372,16 +451,29 @@ async fn view_state(
...
-        fetch_state_from_db(&data.db_manager, account_id, block.block_height, prefix)
-            .await
-            .map_err(|_err| {
-                near_jsonrpc_primitives::types::query::RpcQueryError::UnknownAccount {
-                    requested_account_id: account_id.clone(),
-                    block_height: block.block_height,
-                    block_hash: block.block_hash,
-                }
-            })?;
+    let optimistic_data = data
```
Comment (khorolets) [inline rpc-server/src/modules/queries/methods.rs:454]: If I requested Final, this piece anyway will return me optimistic.

File: rpc-server/src/modules/queries/utils.rs [hunk 5]
```diff
@@ -115,16 +119,36 @@ pub async fn fetch_state_from_db(
         prefix,
     );
     let state_from_db = get_state_keys_from_db(db_manager, account_id, block_height, prefix).await;
-    if state_from_db.is_empty() {
+    if state_from_db.is_empty() && optimistic_data.is_empty() {
```
Comment (khorolets) [inline rpc-server/src/modules/queries/utils.rs:122]: I wouldn't say I like the idea of putting this logic here. I agree to have it here for now but consider adding a TODO to refactor it later.

**Review by khorolets — CHANGES_REQUESTED** [review]
👍 Looks great! We agreed about some fixes and corrections to be done

**Review by khorolets — APPROVED** [review]
Looks good enough for me! Let's start planning the roll-up 👍

**Pull request merged**